# CONTENTS

### 01) Import Libraries and Dataframe
### 02) Wrangling
### 03) Cleaning
### 04) Consistency Checks
### 05) Export Data

# 01 Import Libraries and Dataframe

In [1]:
#Import Libraries
import pandas as pd
import numpy as np
import os

In [2]:
#Settings
pd.set_option('display.max_rows', 500)
pd.set_option('display.max_columns', 500)
pd.set_option('display.width', 1000)

In [3]:
#Import Dataframe
path = r'/Users/Nick/Documents/CareerFoundry/Immersion-6/College-Equity/02-Data'
ced = pd.read_csv(os.path.join(path, 'college-equity-original.csv'), index_col = False)
ced.head()

name                           tier      ownership_type state           city         zip  unitid    opeid  opeid6  student_body  fraction_students_men  fraction_students_women  fraction_students_white  fraction_students_black  fraction_students_hispanic  fraction_students_asian  fraction_students_native_american  fraction_students_pacific_islander  fraction_students_multiracial  fraction_students_non_resident  fraction_students_unknown_ethnicity  fraction_parent_income_q1  fraction_parent_income_q2  fraction_parent_income_q3  fraction_parent_income_q4  fraction_parent_income_q5  mean_parent_income  median_parent_income  fraction_child_income_q1  fraction_child_income_q2  fraction_child_income_q3  fraction_child_income_q4  fraction_child_income_q5  mean_child_income  median_child_income  percent_child_zero_income  mobility_rate  sat_verbal_25_percentile  sat_verbal_75_percentile  sat_math_25_percentile  sat_math_75_percentile  sat_verbal_median  \
0                           ASA College            Two-year for-profit  Private for-profit    NY       Brooklyn  11201-9805  404994  3095500   30955        4437.0                 0.3678                   0.6322                   0.0471                   0.3315                      0.3980                   0.0775                             0.0009                              0.0014                         0.0133                          0.1285                               0.0018                   0.443575                   0.326790                   0.143711                   0.059707                   0.026217         35390.39680                 29000                  0.171280                  0.297217                  0.251632                  0.210021                  0.069851        24916.14900                19700                   0.160623       0.020034                       NaN                       NaN                     NaN                     NaN                NaN   
1          Abilene Christian University              Selective private   Private nonprofit    TX        Abilene       79699  222178   353700    3537        3572.0                 0.4236                   0.5764                   0.6739                   0.0798                      0.1414                   0.0090                             0.0039                              0.0000                         0.0454                          0.0423                               0.0045                   0.052441                   0.101032                   0.163936                   0.258240                   0.424351        138760.96980                101000                  0.140065                  0.132948                  0.173277                  0.246444                  0.307266        50874.14613                40100                   0.121246       0.014364                     470.0                     590.0                   490.0                   600.0              530.0   
2  Abraham Baldwin Agricultural College  Nonselective four-year public              Public    GA         Tifton  31793-2601  138558   154100    1541        3244.0                 0.4750                   0.5250                   0.7830                   0.1239                      0.0586                   0.0071                             0.0025                              0.0003                         0.0092                          0.0151                               0.0003                   0.154555                   0.187159                   0.225497                   0.236763                   0.196027         80366.66127                 66000                  0.121950                  0.179027                  0.255329                  0.273957                  0.169737        35346.92997                32500                   0.113069       0.014857                     410.0                     520.0                   410.0                   510.0              465.0   
3             Academy of Art University           F

# 02 Wrangling

## 02.01 Drop Columns

In [4]:
#Drop columns part 1
drop_list = ['opeid', 'opeid6', 'sat_verbal_25_percentile'
             , 'sat_verbal_75_percentile', 'sat_math_25_percentile'
             , 'sat_math_75_percentile', 'sat_verbal_median'
             , 'sat_math_median', 'sat_equivalent_avg_all']

ced = ced.drop(columns = drop_list)
ced.head()

name                           tier      ownership_type state           city         zip  unitid  student_body  fraction_students_men  fraction_students_women  fraction_students_white  fraction_students_black  fraction_students_hispanic  fraction_students_asian  fraction_students_native_american  fraction_students_pacific_islander  fraction_students_multiracial  fraction_students_non_resident  fraction_students_unknown_ethnicity  fraction_parent_income_q1  fraction_parent_income_q2  fraction_parent_income_q3  fraction_parent_income_q4  fraction_parent_income_q5  mean_parent_income  median_parent_income  fraction_child_income_q1  fraction_child_income_q2  fraction_child_income_q3  fraction_child_income_q4  fraction_child_income_q5  mean_child_income  median_child_income  percent_child_zero_income  mobility_rate  tuition_in_state  tuition_out_of_state  tuition_net  cost_of_attendance  pell_grant_rate  pell_aid_per_student  institutional_grant_aid_avg  \
0                           ASA College            Two-year for-profit  Private for-profit    NY       Brooklyn  11201-9805  404994        4437.0                 0.3678                   0.6322                   0.0471                   0.3315                      0.3980                   0.0775                             0.0009                              0.0014                         0.0133                          0.1285                               0.0018                   0.443575                   0.326790                   0.143711                   0.059707                   0.026217         35390.39680                 29000                  0.171280                  0.297217                  0.251632                  0.210021                  0.069851        24916.14900                19700                   0.160623       0.020034           12298.0               12298.0      14948.0             26546.0           0.7992             3578.4820                      3218.60   
1          Abilene Christian University              Selective private   Private nonprofit    TX        Abilene       79699  222178        3572.0                 0.4236                   0.5764                   0.6739                   0.0798                      0.1414                   0.0090                             0.0039                              0.0000                         0.0454                          0.0423                               0.0045                   0.052441                   0.101032                   0.163936                   0.258240                   0.424351        138760.96980                101000                  0.140065                  0.132948                  0.173277                  0.246444                  0.307266        50874.14613                40100                   0.121246       0.014364           29450.0               29450.0      13974.0             41606.0           0.2595              940.7532                     11832.39   
2  Abraham Baldwin Agricultural College  Nonselective four-year public              Public    GA         Tifton  31793-2601  138558        3244.0                 0.4750                   0.5250                   0.7830                   0.1239                      0.0586                   0.0071                             0.0025                              0.0003                         0.0092                          0.0151                               0.0003                   0.154555                   0.187159                   0.225497                   0.236763                   0.196027         80366.66127                 66000                  0.121950                  0.179027                  0.255329                  0.273957                  0.169737        35346.92997                32500                   0.113069       0.014857            3394.0                9839.0       2359.0             11674.0           0.4629             1870.4520                      1152.64   
3             Academy of Art Un

In [5]:
#Drop columns part 2
#Drop last N columns (all fraction_majors_X columns)
N = 39
ced = ced.iloc[:, :-N]
ced.head()

name                           tier      ownership_type state           city         zip  unitid  student_body  fraction_students_men  fraction_students_women  fraction_students_white  fraction_students_black  fraction_students_hispanic  fraction_students_asian  fraction_students_native_american  fraction_students_pacific_islander  fraction_students_multiracial  fraction_students_non_resident  fraction_students_unknown_ethnicity  fraction_parent_income_q1  fraction_parent_income_q2  fraction_parent_income_q3  fraction_parent_income_q4  fraction_parent_income_q5  mean_parent_income  median_parent_income  fraction_child_income_q1  fraction_child_income_q2  fraction_child_income_q3  fraction_child_income_q4  fraction_child_income_q5  mean_child_income  median_child_income  percent_child_zero_income  mobility_rate  tuition_in_state  tuition_out_of_state  tuition_net  cost_of_attendance  pell_grant_rate  pell_aid_per_student  institutional_grant_aid_avg  \
0                           ASA College            Two-year for-profit  Private for-profit    NY       Brooklyn  11201-9805  404994        4437.0                 0.3678                   0.6322                   0.0471                   0.3315                      0.3980                   0.0775                             0.0009                              0.0014                         0.0133                          0.1285                               0.0018                   0.443575                   0.326790                   0.143711                   0.059707                   0.026217         35390.39680                 29000                  0.171280                  0.297217                  0.251632                  0.210021                  0.069851        24916.14900                19700                   0.160623       0.020034           12298.0               12298.0      14948.0             26546.0           0.7992             3578.4820                      3218.60   
1          Abilene Christian University              Selective private   Private nonprofit    TX        Abilene       79699  222178        3572.0                 0.4236                   0.5764                   0.6739                   0.0798                      0.1414                   0.0090                             0.0039                              0.0000                         0.0454                          0.0423                               0.0045                   0.052441                   0.101032                   0.163936                   0.258240                   0.424351        138760.96980                101000                  0.140065                  0.132948                  0.173277                  0.246444                  0.307266        50874.14613                40100                   0.121246       0.014364           29450.0               29450.0      13974.0             41606.0           0.2595              940.7532                     11832.39   
2  Abraham Baldwin Agricultural College  Nonselective four-year public              Public    GA         Tifton  31793-2601  138558        3244.0                 0.4750                   0.5250                   0.7830                   0.1239                      0.0586                   0.0071                             0.0025                              0.0003                         0.0092                          0.0151                               0.0003                   0.154555                   0.187159                   0.225497                   0.236763                   0.196027         80366.66127                 66000                  0.121950                  0.179027                  0.255329                  0.273957                  0.169737        35346.92997                32500                   0.113069       0.014857            3394.0                9839.0       2359.0             11674.0           0.4629             1870.4520                      1152.64   
3             Academy of Art Un

# 03 Cleaning

## 03.01 Nulls and Data Types

In [6]:
ced.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3645 entries, 0 to 3644
Data columns (total 69 columns):
 #   Column                                 Non-Null Count  Dtype  
---  ------                                 --------------  -----  
 0   name                                   3645 non-null   object 
 1   tier                                   3645 non-null   object 
 2   ownership_type                         3645 non-null   object 
 3   state                                  3645 non-null   object 
 4   city                                   3645 non-null   object 
 5   zip                                    3645 non-null   object 
 6   unitid                                 3645 non-null   int64  
 7   student_body                           3351 non-null   float64
 8   fraction_students_men                  3351 non-null   float64
 9   fraction_students_women                3351 non-null   float64
 10  fraction_students_white                3351 non-null   float64
 11  frac

#### Due to the nature of the variables, we can infer that all null values represent missing values--no other significance. There are 3645 entries, so any variable with less than 2916 non-null values (729 or more nulls) would have 20% or more nulls. Those variables should be dropped. Unfortunately, this throws out all the demographic completion rate variables.

In [7]:
#Drop columns with >= 20% nulls (729 nulls)
ced = ced[[col for col in ced if ced[col].isnull().sum() <= 729]]
ced.head()

name                           tier      ownership_type state           city         zip  unitid  student_body  fraction_students_men  fraction_students_women  fraction_students_white  fraction_students_black  fraction_students_hispanic  fraction_students_asian  fraction_students_native_american  fraction_students_pacific_islander  fraction_students_multiracial  fraction_students_non_resident  fraction_students_unknown_ethnicity  fraction_parent_income_q1  fraction_parent_income_q2  fraction_parent_income_q3  fraction_parent_income_q4  fraction_parent_income_q5  mean_parent_income  median_parent_income  fraction_child_income_q1  fraction_child_income_q2  fraction_child_income_q3  fraction_child_income_q4  fraction_child_income_q5  mean_child_income  median_child_income  percent_child_zero_income  mobility_rate  tuition_in_state  tuition_out_of_state  tuition_net  cost_of_attendance  pell_grant_rate  pell_aid_per_student  institutional_grant_aid_avg  \
0                           ASA College            Two-year for-profit  Private for-profit    NY       Brooklyn  11201-9805  404994        4437.0                 0.3678                   0.6322                   0.0471                   0.3315                      0.3980                   0.0775                             0.0009                              0.0014                         0.0133                          0.1285                               0.0018                   0.443575                   0.326790                   0.143711                   0.059707                   0.026217         35390.39680                 29000                  0.171280                  0.297217                  0.251632                  0.210021                  0.069851        24916.14900                19700                   0.160623       0.020034           12298.0               12298.0      14948.0             26546.0           0.7992             3578.4820                      3218.60   
1          Abilene Christian University              Selective private   Private nonprofit    TX        Abilene       79699  222178        3572.0                 0.4236                   0.5764                   0.6739                   0.0798                      0.1414                   0.0090                             0.0039                              0.0000                         0.0454                          0.0423                               0.0045                   0.052441                   0.101032                   0.163936                   0.258240                   0.424351        138760.96980                101000                  0.140065                  0.132948                  0.173277                  0.246444                  0.307266        50874.14613                40100                   0.121246       0.014364           29450.0               29450.0      13974.0             41606.0           0.2595              940.7532                     11832.39   
2  Abraham Baldwin Agricultural College  Nonselective four-year public              Public    GA         Tifton  31793-2601  138558        3244.0                 0.4750                   0.5250                   0.7830                   0.1239                      0.0586                   0.0071                             0.0025                              0.0003                         0.0092                          0.0151                               0.0003                   0.154555                   0.187159                   0.225497                   0.236763                   0.196027         80366.66127                 66000                  0.121950                  0.179027                  0.255329                  0.273957                  0.169737        35346.92997                32500                   0.113069       0.014857            3394.0                9839.0       2359.0             11674.0           0.4629             1870.4520                      1152.64   
3             Academy of Art Un

In [8]:
ced.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3645 entries, 0 to 3644
Data columns (total 60 columns):
 #   Column                                 Non-Null Count  Dtype  
---  ------                                 --------------  -----  
 0   name                                   3645 non-null   object 
 1   tier                                   3645 non-null   object 
 2   ownership_type                         3645 non-null   object 
 3   state                                  3645 non-null   object 
 4   city                                   3645 non-null   object 
 5   zip                                    3645 non-null   object 
 6   unitid                                 3645 non-null   int64  
 7   student_body                           3351 non-null   float64
 8   fraction_students_men                  3351 non-null   float64
 9   fraction_students_women                3351 non-null   float64
 10  fraction_students_white                3351 non-null   float64
 11  frac

In [9]:
#Change unitid from int to string
ced['unitid'] = ced['unitid'].astype('str')
ced['unitid'].dtype

dtype('O')

In [10]:
#Check for mixed-type data
for col in ced.columns.tolist():
    weird = (ced[[col]].applymap(type) 
             != ced[[col]].iloc[0].apply(type)).any(axis = 1)
    if len(ced[weird]) > 0:
        print(col)

#### No columns with mixed-type data

## 03.02  Duplicates

In [11]:
#Check original dimensions
ced.shape

(3645, 60)

In [12]:
#Check for duplicates
ced[ced.duplicated()]

Empty DataFrame
Columns: [name, tier, ownership_type, state, city, zip, unitid, student_body, fraction_students_men, fraction_students_women, fraction_students_white, fraction_students_black, fraction_students_hispanic, fraction_students_asian, fraction_students_native_american, fraction_students_pacific_islander, fraction_students_multiracial, fraction_students_non_resident, fraction_students_unknown_ethnicity, fraction_parent_income_q1, fraction_parent_income_q2, fraction_parent_income_q3, fraction_parent_income_q4, fraction_parent_income_q5, mean_parent_income, median_parent_income, fraction_child_income_q1, fraction_child_income_q2, fraction_child_income_q3, fraction_child_income_q4, fraction_child_income_q5, mean_child_income, median_child_income, percent_child_zero_income, mobility_rate, tuition_in_state, tuition_out_of_state, tuition_net, cost_of_attendance, pell_grant_rate, pell_aid_per_student, institutional_grant_aid_avg, fraction_students_with_debt, debt_median, debt_90_percentile, debt_75_percentile, debt_25_percentile, debt_10_percentile, online_only, completion_rate, student_income_6_years_mean, student_income_6_years_median, student_income_6_years_25_percentile, student_income_6_years_75_percentile, student_income_6_years_stdev, student_income_10_years_mean, student_income_10_years_median, student_income_10_years_25_percentile, student_income_10_years_75_percentile, student_income_10_years_stdev]
Index: []

#### No duplicates

# 04 Consistency Checks

In [13]:
#Summary stats for each variable
ced.describe()

student_body  fraction_students_men  fraction_students_women  fraction_students_white  fraction_students_black  fraction_students_hispanic  fraction_students_asian  fraction_students_native_american  fraction_students_pacific_islander  fraction_students_multiracial  fraction_students_non_resident  fraction_students_unknown_ethnicity  fraction_parent_income_q1  fraction_parent_income_q2  fraction_parent_income_q3  fraction_parent_income_q4  fraction_parent_income_q5  mean_parent_income  median_parent_income  fraction_child_income_q1  fraction_child_income_q2  fraction_child_income_q3  fraction_child_income_q4  fraction_child_income_q5  mean_child_income  median_child_income  percent_child_zero_income  mobility_rate  tuition_in_state  tuition_out_of_state    tuition_net  cost_of_attendance  pell_grant_rate  pell_aid_per_student  institutional_grant_aid_avg  fraction_students_with_debt   debt_median  debt_90_percentile  debt_75_percentile  debt_25_percentile  debt_10_percentile  \
count    3351.000000            3351.000000              3351.000000              3351.000000              3351.000000                 3351.000000              3351.000000                        3351.000000                         3351.000000                    3351.000000                     3351.000000                          3351.000000                3645.000000                3645.000000                3645.000000                3645.000000                3645.000000         3645.000000           3645.000000               3645.000000               3645.000000               3645.000000               3645.000000               3645.000000        3645.000000          3645.000000                3645.000000    3645.000000       3084.000000           3421.000000    3367.000000         3284.000000      3346.000000           3280.000000                  3302.000000                  3297.000000   3531.000000         3509.000000         3522.000000         3522.000000         3509.000000   
mean     4311.747240               0.409315                 0.590685                 0.532039                 0.173343                    0.135125                 0.035066                           0.010236                            0.003285                       0.029105                        0.020975                             0.060827                   0.145443                   0.182251                   0.211805                   0.233559                   0.226943        96799.751972          71238.463649                  0.139513                  0.183334                  0.216806                  0.233044                  0.227303       39338.913900         34088.724280                   0.126164       0.018312      14552.291505          17957.195557   10237.526582        24960.092266         0.472021           1817.201846                  4155.178680                     0.572745  11472.520674        27926.962382        19853.228279         5173.865985         2845.701482   
std      6987.193475               0.175365                 0.175365                 0.244394                 0.194397                    0.161840                 0.052260                           0.031064                            0.014763                       0.030185                        0.038580                             0.104303                   0.094636                   0.068130                   0.046851                   0.063060                   0.149887        57671.938361          26680.270009                  0.048665                  0.071917                  0.060654                  0.054834                  0.138353       16124.416191         12553.223754                   0.043557       0.013033      12095.845122          10666.977375    7667.519491        12993.947959         0.204684            833.113617                  5564.553184                     0.291976   5594.198614        10340.786000         7838.577197         2410.384487         1037.585575   
min 

#### Things to investigate: college(s) with 1 student, college(s) with tuition of 0

In [14]:
#Colleges with 1 student
ced_1stud = ced[ced['student_body'] == 1]
ced_1stud

name                  tier      ownership_type state          city    zip  unitid  student_body  fraction_students_men  fraction_students_women  fraction_students_white  fraction_students_black  fraction_students_hispanic  fraction_students_asian  fraction_students_native_american  fraction_students_pacific_islander  fraction_students_multiracial  fraction_students_non_resident  fraction_students_unknown_ethnicity  fraction_parent_income_q1  fraction_parent_income_q2  fraction_parent_income_q3  fraction_parent_income_q4  fraction_parent_income_q5  mean_parent_income  median_parent_income  fraction_child_income_q1  fraction_child_income_q2  fraction_child_income_q3  fraction_child_income_q4  fraction_child_income_q5  mean_child_income  median_child_income  percent_child_zero_income  mobility_rate  tuition_in_state  tuition_out_of_state  tuition_net  cost_of_attendance  pell_grant_rate  pell_aid_per_student  institutional_grant_aid_avg  \
1310  ITT Technical Institute-Cedar Rapids  Four-year for-profit  Private for-profit    IA  Cedar Rapids  52404  459620           1.0                    1.0                      0.0                      1.0                      0.0                         0.0                      0.0                                0.0                                 0.0                            0.0                             0.0                                  0.0                   0.194546                   0.241633                   0.239987                   0.209592                   0.114243         65233.37011                 52600                  0.127305                   0.15362                  0.213984                  0.270258                  0.234832        38044.73506                35600                   0.116113       0.036342               NaN                   NaN      48680.0                 NaN           0.5676              2291.695                       1110.0   

      fraction_students_with_debt  debt_median  debt_90_percentile  debt_75_percentile  debt_25_percentile  debt_10_percentile  online_only  completion_rate  student_income_6_years_mean  student_income_6_years_median  student_income_6_years_25_percentile  student_income_6_years_75_percentile  student_income_6_years_stdev  student_income_10_years_mean  student_income_10_years_median  student_income_10_years_25_percentile  student_income_10_years_75_percentile  student_income_10_years_stdev  
1310                          NaN      12297.5             33263.0             25000.0              3500.0              2768.0          0.0              NaN                      34800.0                        31400.0                               16800.0                               48000.0                       25200.0                       44000.0                         39400.0                                22500.0                                58100.0                        32700.0

#### No indication as to why this school only has one student. No action to take.

In [15]:
#Colleges with tuition of 0
ced_0tuit = ced[ced['tuition_in_state'] == 0]
ced_0tuit

name                  tier      ownership_type state      city        zip    unitid  student_body  fraction_students_men  fraction_students_women  fraction_students_white  fraction_students_black  fraction_students_hispanic  fraction_students_asian  fraction_students_native_american  fraction_students_pacific_islander  fraction_students_multiracial  fraction_students_non_resident  fraction_students_unknown_ethnicity  fraction_parent_income_q1  fraction_parent_income_q2  fraction_parent_income_q3  fraction_parent_income_q4  fraction_parent_income_q5  mean_parent_income  median_parent_income  fraction_child_income_q1  fraction_child_income_q2  fraction_child_income_q3  fraction_child_income_q4  fraction_child_income_q5  mean_child_income  median_child_income  percent_child_zero_income  mobility_rate  tuition_in_state  tuition_out_of_state  tuition_net  cost_of_attendance  pell_grant_rate  pell_aid_per_student  institutional_grant_aid_avg  \
1619  Lincoln College of New England-Hartford  Four-year for-profit  Private for-profit    CT  Hartford  061053630  12868301           NaN                    NaN                      NaN                      NaN                      NaN                         NaN                      NaN                                NaN                                 NaN                            NaN                             NaN                                  NaN                   0.135511                   0.163028                   0.202264                   0.259153                   0.240044         80374.39154                 75000                  0.112286                  0.183589                  0.300253                  0.280049                  0.123822        29802.19031                29100                   0.107277       0.006345               0.0                   NaN          NaN                 NaN              NaN                   NaN                          NaN   

      fraction_students_with_debt  debt_median  debt_90_percentile  debt_75_percentile  debt_25_percentile  debt_10_percentile  online_only  completion_rate  student_income_6_years_mean  student_income_6_years_median  student_income_6_years_25_percentile  student_income_6_years_75_percentile  student_income_6_years_stdev  student_income_10_years_mean  student_income_10_years_median  student_income_10_years_25_percentile  student_income_10_years_75_percentile  student_income_10_years_stdev  
1619                          NaN      12000.0             35750.0             25199.0              5250.0              3500.0          NaN              NaN                      31800.0                        29200.0                               14500.0                               42600.0                       22600.0                       36200.0                         33900.0                                19900.0                                49900.0                        22600.0

#### This college only has student income data and basic geographic info, 0 seems to be a replacement for null here. I'll keep this observation as it will contribute to any income-related analysis.

# 05 Export Data

In [17]:
ced.to_csv(os.path.join(path, 'college-equity-clean.csv'))